### PR TITLE
Support React Native Windows - C++/WinRT and C# templates with Example

### LINKED ISSUE #69: Add template for React Native Windows plugins
# Feature request

It would be nice to have a plugin template for [React Native Windows](https://github.com/microsoft/react-native-windows). Perhaps someone from Microsoft (@stmoy?) might be able to help out?

## Why it is needed

Adding a template for RN Windows will help potential contributors follow a "best-practice" way of adding support for Windows to their desired plugins

## Possible implementation

Windows projects use Visual Studio solution files (`*.sln`). The following repositories have Windows support:

- [`react-native-webview`](https://github.com/react-native-community/react-native-webview)
- [`async-storage`](https://github.com/react-native-community/async-storage)
- [`react-native-netinfo`](https://github.com/react-native-community/react-native-netinfo)

### PR DESCRIPTION
**Pull request by BartoszKlonowski**
This pull request closes #69 

It delivers the full support of React Native Windows project's templates for both C# and C++ languages.
Example project is configured as C# application.

---

Following changes were implemented:
* Additional questions and options in the *index.ts* file for Windows support
* *package.json* extended with *react-native-windows: v0.63.4*
* C++/WinRT project templates for both native module and native UI component
* C# project templates for both native module and native UI component
* README updated with the information about Windows support
* Example application provided with the Windows solution (*.sln*)

The 4 new project templates created in the Bob project are created as a 0.63.x projects with React Native Windows dependency already added to the solution configuration.
None of the scaffolded projects use the precompiled headers. All headers required for implementation are included where they are required.

**NOTE:** Each language has its project's GUID being the same across native module and native UI component.

All projects are provided with the most basic source code file, which should be a starting point for a further implementation.
Though they are basic and do nothing, they still compile successfully for each project.
Those files and classes they implement are named according to the type of project selected.
These basic modules/component are already registered in the *ReactPackageProvider*.

---

Please see the short presentation of the full template creation looks like (C# is taken as an example, but the procedure is of course the same for C++/WinRT):

BobExample

---

For some more explanations please check the commit messages.

### CODE REVIEW

**Comment by asklar** [comment]
Love seeing the support for Windows apps being added. Main piece of feedback is that the project files and the code that the template RNW CLI produce, change over time. That is to say, an app created for 0.63 RNW will have different project/code files as 0.64 RNW. We expect we will continue to iterate on the template code and project files as we continue to evolve the platform. Sometimes we also update the template files/projects to backport a severe bugfix.

So ideally react-native-builder-bob wouldn't carry these files itself and instead just call the RNW CLI (or somehow "suck up" the files that RNW CLI uses for each version).

**Comment by asklar** [comment]
@jonthysell to review too

**Comment by BartoszKlonowski** [comment]
@asklar You've got the point - creating the project through the CLI would prevent us from updating the template and track backports. On the other hand current solution allows to do some customizations to the scaffolded solution, also whole system (each language, platform) is done like this so there's no confusion for a future maintainer, I also believe that the amount of changes required to be done to implement CLI for Windows only would be bigger than the current solution.
Of course, the other side of a coin is that the XML file can be more confusing than command line, but I'm aware of that and to be honest I'm willing to keep on maintaining that.

**But** I see no problem in keeping your opinion in mind, so perhaps whole system will be based on CLI, not just Windows.

**Comment by asklar** [comment]
Yes I can see your point; I wouldn't want Windows to be the odd one out that's different : ) I just worry about how RNBB will be able to keep up/know about changes to the templates, but if you guys are comfortable signing up to keep this maintained then I am not concerned :)

**Comment by BartoszKlonowski** [comment]
@asklar What just came to my mind is that there's one main concern about CLI-based approach:
*react-native-windows-init* have the `--projectType` flag which determines whether *exe* or *dll* will be created, but if it's the native module or native UI component is determined only by the files and the implementation of the project. So there would be no possibility to create those two separately.
I'm glad that we agree about the current approach, but CLI-based approach should be something to think about once it's possible for each platfrom and environment.

**Comment by BartoszKlonowski** [comment]
@asklar @jonthysell Did you take a chance to take a look at .vcxproj and .csproj files? These are essential for the Builder Bob tool to work correctly and your expertise is most valuable here.

File: packages/create-react-native-library/src/index.ts
Comment (asklar) [inline packages/create-react-native-library/src/index.ts:391]: suggest that we call these uwp-cpp and uwp-csharp, or windows-uwp-cpp / csharp
We will eventually have non-UWP templates too (working on it as we speak), which are for win32 apps using WinUI 3 and the Windows App SDK (formerly known as Project Reunion)

File: packages/create-react-native-library/templates/example/example/windows/{%- project.name %}Example/App.xaml.cs
Comment (asklar) [inline packages/create-react-native-library/templates/example/example/windows/{%- project.name %}Example/App.xaml.cs:30]: I think we recently deleted this line (this past week in fact :) )

File: packages/create-react-native-library/templates/example/example/windows/{%- project.name %}Example/Package.appxmanifest
Comment (asklar) [inline packages/create-react-native-library/templates/example/example/windows/{%- project.name %}Example/Package.appxmanifest:11]: this should probably be the user's name or something in the template :)

File: packages/create-react-native-library/templates/windows-cpp-library/windows/{%- project.name %}/readme.txt
Comment (asklar) [inline packages/create-react-native-library/templates/windows-cpp-library/windows/{%- project.name %}/readme.txt:1]: nit, can delete this file

File: packages/create-react-native-library/templates/windows-cpp-view-library/windows/{%- project.name %}/readme.txt
Comment (asklar) [inline packages/create-react-native-library/templates/windows-cpp-view-library/windows/{%- project.name %}/readme.txt:1]: can delete and remove this file from the project too :)

File: packages/create-react-native-library/templates/windows-cpp-view-library/windows/{%- project.name %}/{%- project.name %}ViewManager.hpp
Comment (asklar) [inline packages/create-react-native-library/templates/windows-cpp-view-library/windows/{%- project.name %}/{%- project.name %}ViewManager.hpp:11]: I think you'll probably want to implement some of the other interfaces, e.g. [IViewManagerWithReactContext](https://microsoft.github.io/react-native-windows/docs/IViewManagerWithReactContext)

**Review by asklar — APPROVED** [review]
Recommend adding CI validation that the apps/libs/modules can be generated properly as well as that they build :)

File: packages/create-react-native-library/templates/windows-cpp-view-library/windows/{%- project.name %}/{%- project.name %}ViewManager.hpp
Comment (BartoszKlonowski) [reply packages/create-react-native-library/templates/windows-cpp-view-library/windows/{%- project.name %}/{%- project.name %}ViewManager.hpp:11]: I wanted to keep the starting point simple so a developer can add whatever will be used, but if you recommend adding this then I'll go with your advice 👍

File: packages/create-react-native-library/templates/example/example/windows/{%- project.name %}Example/Package.appxmanifest
Comment (BartoszKlonowski) [reply packages/create-react-native-library/templates/example/example/windows/{%- project.name %}Example/Package.appxmanifest:11]: I completely forgot about this, thanks!
I will double check whether there are more of such mistakes.

File: packages/create-react-native-library/templates/example/example/windows/{%- project.name %}Example/App.xaml.cs
Comment (BartoszKlonowski) [reply packages/create-react-native-library/templates/example/example/windows/{%- project.name %}Example/App.xaml.cs:30]: This template is for late of v0.63, so I will need to keep this line, but once the whole Bob is updated, I'll go with removing it 👍

**Comment by jahead** [comment]
What's the state of this?

**Comment by BartoszKlonowski** [comment]
@jahead We are still waiting for the review and will only merge this with an approval of an owner: @satya164

**Comment by jahead** [comment]
@satya164 do you have some time for you to review this?

**Comment by satya164** [comment]
@BartoszKlonowski can you rebase this PR on top of main if you get some time this week? then I'll have time to review and merge this.

**Comment by rtorr** [comment]
looking forward to this

**Comment by satya164** [comment]
Closing due to inactivity. I'd like to figure out a different way to support other platforms without adding the code to this repo.

**Pull request closed without merging**
